### PR TITLE
Android R8 fixes to enable optimizations

### PR DESCRIPTION
**Pull request by biodranik**
@pastk Надо допилить этот PR, хорошо его тестируя на `./gradlew runFDroidBeta`

### CODE REVIEW

File: android/src/com/mapswithme/maps/MwmActivity.java [hunk 1]
```diff
@@ -1118,7 +1117,7 @@ private boolean removeFragment(String className, boolean animate)
   }
 
   // Called from JNI.
-  @Override
+  @Override @Keep
```
[the remark this answers is not in the document]
Comment (biodranik) [reply android/src/com/mapswithme/maps/MwmActivity.java:1120]: I don't like it, but ok. Can you help?

File: android/proguard-mwm.txt [hunk 2]
```diff
@@ -1,27 +1,2 @@
...
-}
-
--dontwarn com.google.android.material.**
--keep class com.google.android.material.** { *; }
-
--dontwarn androidx.**
--keep class androidx.** { *; }
--keep interface androidx.** { *; }
+# Every JNI-called class and method should have a @Keep annotation to properly work and to avoid
+# being manually added as -keep in this file.
```
[the remark this answers is not in the document]
Comment (biodranik) [reply android/proguard-mwm.txt:2]: Can you please help?

File: android/src/com/mapswithme/maps/api/ParsedSearchRequest.java [hunk 3]
```diff
@@ -1,13 +1,14 @@
...
 
+import androidx.annotation.Keep;
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 
 /**
  * Represents url_scheme::SearchRequest from core.
  */
-
+@Keep
```
[the remark this answers is not in the document]
Comment (biodranik) [reply android/src/com/mapswithme/maps/api/ParsedSearchRequest.java:11]: We don't use reflection, so `@Keep` is the obvious way to mark all JNI calls. The KISS principle.

File: android/src/com/mapswithme/maps/MwmActivity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pastk) [reply android/src/com/mapswithme/maps/MwmActivity.java:1120]: You can use "Reformat file" - "Only VCS changed lines" in AS.

File: android/proguard-mwm.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (pastk) [reply android/proguard-mwm.txt:2]:
```suggestion
# being manually added as -keep in this file, for details see https://developer.android.com/studio/build/shrink-code#keep-code
```

**Comment by pastk** [comment]
The app starts now, but needs more testing to fix more crashes.

Fdroid beta x86_64 apk is reduced by 1.7Mb :)

**Comment by biodranik** [comment]
Do you have actions access to run the android beta on your branch?

File: android/proguard-mwm.txt [hunk 4]
```diff
@@ -1,38 +1,9 @@
-# obfuscate supportV7 menu subclasses. it fixes bug with some Samsung and other devices ROMS based on android 4.2.2.
-# more details here : https://code.google.com/p/android/issues/detail?id=78377
-# For some reason, this line disables optimizations and avoids crashes due to missing @Keep attributes.
-# Looks like R8 keeps EVERYTHING except that unused support lib :)
-# TODO: Remove this line after properly marking all JNI-called classes and methods with @Keep.
-# Also remove everything else what is not needed.
--keep class !android.support.v7.internal.view.menu.**,** {*;}
+# Every JNI-called class and method should have a @Keep annotation to properly work and to avoid
```
Comment (rtsisyk) [inline android/proguard-mwm.txt:1]: What is the purpose of this patch?

Comment (biodranik) [reply android/proguard-mwm.txt:1]: To enable Java code optimizations and shrinking.

File: 3party/opening_hours/opening_hours.cpp [hunk 5]
```diff
@@ -22,9 +22,16 @@
   THE SOFTWARE.
 */
 
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-copy"
```
Comment (rtsisyk) [inline 3party/opening_hours/opening_hours.cpp:27]: Why is this warning suppressed? Is it related to this patch?

File: android/proguard-mwm.txt [hunk 6]
```diff
@@ -1,38 +1,9 @@
-# obfuscate supportV7 menu subclasses. it fixes bug with some Samsung and other devices ROMS based on android 4.2.2.
-# more details here : https://code.google.com/p/android/issues/detail?id=78377
-# For some reason, this line disables optimizations and avoids crashes due to missing @Keep attributes.
-# Looks like R8 keeps EVERYTHING except that unused support lib :)
-# TODO: Remove this line after properly marking all JNI-called classes and methods with @Keep.
-# Also remove everything else what is not needed.
--keep class !android.support.v7.internal.view.menu.**,** {*;}
```
Comment (rtsisyk) [inline android/proguard-mwm.txt:7]: This was an old support library we don't use anymore.

File: android/src/com/mapswithme/maps/MwmActivity.java [hunk 7]
```diff
@@ -1115,7 +1117,7 @@ private boolean removeFragment(String className, boolean animate)
...
```
Comment (rtsisyk) [inline android/src/com/mapswithme/maps/MwmActivity.java:1120]: Could you please keep annotations on new lines to follow existing convention in the code?

File: android/src/com/mapswithme/maps/bookmarks/data/Bookmark.java [hunk 8]
```diff
@@ -13,7 +14,7 @@
 import com.mapswithme.util.Constants;
 
 // TODO consider refactoring to remove hack with MapObject unmarshalling itself and Bookmark at the same time.
-@SuppressLint("ParcelCreator")
+@SuppressLint("ParcelCreator") @Keep
```
Comment (rtsisyk) [inline android/src/com/mapswithme/maps/bookmarks/data/Bookmark.java:17]: keep annotations on a separate line

File: android/proguard-mwm.txt [hunk 9]
```diff
@@ -1,38 +1,9 @@
...
-
--keepclassmembers class * implements com.mapswithme.util.Gsonable {
-  <init>(...);
-}
-
--dontwarn com.google.android.material.**
--keep class com.google.android.material.** { *; }
-
--dontwarn androidx.**
--keep class androidx.** { *; }
```
Comment (rtsisyk) [inline android/proguard-mwm.txt:37]: Are you sure here? This namespace is still used by app.

**Comment by rtsisyk** [comment]
Replaced by https://github.com/organicmaps/organicmaps/pull/6159

**Pull request closed without merging**